apps/docusaurus/docs/guides/randomness.md

```diff
@@ -0,0 +1,314 @@
+
+```rust
+module module_owner::lottery {
+    // ...
+
+    #[randomness]
+    entry fun decide_winner_v0() {
+        // ...
+        decide_winner_internal(lottery_state);
+    }
+
+    #[randomness]
+    entry fun decide_winner_v1() {
+        // ...
+        decide_winner_internal(lottery_state);
+    }
+
+    // A private helper function
+    fun decide_winner_internal(lottery_state: &mut lottery_state) {
+        let n = std::vector::length(&lottery_state.players);
+        let winner_idx = aptos_framework::randomness::u64_range(0, n);
+        lottery_state.winner_idx = std::option::some(winner_idx);
+    }
+}
+```
+
+If `decide_winner_internal()` were accidentally marked public,
+malicious players can deploy their own contract to:
+
+1. call`decide_winner_internal()`;
+1. read the lottery result (assuming the `lottery` module has some getter functions for the result);
+1. abort if the result is not in their favor.
+   By repeatedly calling their own contract until a txn succeeds,
+   malicious users can bias the uniform distribution of the winner (dApp developer's initial design).
+   This is referred to as a _test-and-abort attack_.
+
+Aptos move compiler have been updated to prevent this attack for your contract safety:
+a randomness-dependent public function is treated as a compile error.
+If you have finished the steps in the ["build aptos-cli"](#build-aptos-cli-from-latest-source)) section,
+then your aptos-cli are equipped with the updated compiler.
+
+```rust
+module module_owner::lottery {
+    // Compile error!
+    public fun decide_winner_internal(lottery_state: &mut lottery_state) {
+        let n = std::vector::length(&lottery_state.players);
+        let winner_idx = aptos_framework::randomness::u64_range(0, n);
+        lottery_state.winner_idx = std::option::some(winner_idx);
+    }
+}
+```
+
+Not recommended, but if you intend to expose such a randomness-dependent function to the public, you can bypass the compiler check by annotating your function with `#[lint_allow_unsafe_randomness]`.
+
+```rust
+module module_owner::lottery {
+    // Can compile, but use it at your own risk!
+    #[lint_allow_unsafe_randomness]
+    public fun decide_winner_internal(lottery_state: &mut lottery_state) {
+        let n = std::vector::length(&lottery_state.players);
+        let winner_idx = aptos_framework::randomness::u64_range(0, n);
+        lottery_state.winner_idx = std::option::some(winner_idx);
+    }
+}
+```
+
+### Security warning: undergasing attacks, and how to prevent
+
+Imagine such a dApp. It defines a private entry function for a user to:
+
+1. toss a coin (gas cost: 9), then
+2. get a reward (gas cost: 10) if coin=1, or get multiple punishments (gas cost: 100) otherwise.
+
+A malicious user can control its account blanace so it covers at most 108 gas units (or set transaction parameter `max_gas=108`), and the punishing branch will never be reached.
+
+Formally, this is referred to as an _undergasing attack_,
```

Comment:
Link to https://github.com/aptos-foundation/AIPs/blob/main/aips/aip-41.md#undergasing-attacks